icu4c/source/i18n/unicode/ucol.h

```diff
@@ -1519,6 +1519,130 @@ ucol_openBinary(const uint8_t *bin, int32_t length,
                 const UCollator *base, 
                 UErrorCode *status);
 
+#if U_SHOW_CPLUSPLUS_API
+
+#include <functional>
+#include <string_view>
+#include <type_traits>
+
+#include "unicode/char16ptr.h"
+#include "unicode/stringpiece.h"
+#include "unicode/unistr.h"
+
+namespace U_HEADER_ONLY_NAMESPACE {
+
+#ifndef U_HIDE_DRAFT_API
+
+namespace collator {
+
+namespace internal {
+
+/**
+ * Function object for performing comparisons using a UCollator.
+ * @internal
+ */
+template <template <typename...> typename Compare, UCollationResult result>
+class Predicate {
+  public:
+    /** @internal */
+    explicit Predicate(const UCollator* ucol) : collator(ucol) {}
+
+    /** @internal */
+    template <
+        typename T, typename U,
+        typename = std::enable_if_t<ConvertibleToU16StringView<T> && ConvertibleToU16StringView<U>>>
+    bool operator()(const T& lhs, const U& rhs) const {
+        return strcoll(UnicodeString::readOnlyAlias(lhs), UnicodeString::readOnlyAlias(rhs));
+    }
+
+    /** @internal */
+    bool operator()(std::string_view lhs, std::string_view rhs) const {
+        return strcollUTF8(lhs, rhs);
+    }
+
+#if defined(__cpp_char8_t)
+    /** @internal */
+    bool operator()(std::u8string_view lhs, std::u8string_view rhs) const {
+        return strcollUTF8(lhs, rhs);
+    }
+#endif
+
+  private:
+    bool strcoll(UnicodeString lhs, UnicodeString rhs) const {
```

Comment:
(copied from code review comment in API proposal doc)
I know these are internal/private functions, but I got tripped up trying to understand their call sites, and got tripped up because they have the usual names suggesting a three-way result, but these actually also call "compare(..., result)" and return a bool.

I would be clearer to give them new names, like changing strcoll here to... apply? As in, applying the predicate function?

(New addition) The ...UTF8 suffix for those versions is good.

Reply:
In the context of the standard library, "apply" normally means something else. Maybe "match" could be a simple and neutral name?

Reply:
"match" could work.
Or, in keeping with these being thin wrappers over the "compare" predicate, maybe "compareUTF16" and "compareUTF8".

Reply:
> Or, in keeping with these being thin wrappers over the "compare" predicate, maybe "compareUTF16" and "compareUTF8".

Then we'd have two different things named compare, I don't think that'd be an improvement.

Reply:
A minute later I saw your commit changing to "match". Let's move forward with that. Hence approval.